changing lodash

brittanyjoiner15 opened the pull request:
Resolves #73937

Updated all instances of `import { x } from 'lodash'` to `import x from 'lodash/x'`

[Used this codemod](https://github.com/dgrijuela/modular-lodash-codemod)

watson reviewed (approved):
LGTM from a security perspective 👍 

There seem to be a few places missing though, e.g:

https://github.com/elastic/kibana/blob/b3202a0e4cba43a54183f732d3092368f9e63fd1/x-pack/plugins/canvas/public/lib/parse_single_function_chain.js#L7

Have those been left out on purpose?

brittanyjoiner15 commented:
Just realized it's not just the mapValues :) I'll go find the other lodash ones and update and then move out of draft.

brittanyjoiner15 commented:
@watson i realized i only did mapvalues and not all the other lodash ones. I've since updated and ran code mod to do them all!

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/66935/)
* Commit: fbdae3fa444b19f35edcb40e309c5bdc7ecc300c

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/fbdae3fa444b19f35edcb40e309c5bdc7ecc300c) | diff | [baseline](https://github.com/elastic/kibana/commit/c655f50950693f3d0ec32dcf8167135d0c9781ec) |
| --- | --- | --- | --- |
| canvas | 1270 | +67 | 1203 |
| spaces | 312 | +120 | 192 |
| total |  | +187 |  |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/fbdae3fa444b19f35edcb40e309c5bdc7ecc300c) | diff | [baseline](https://github.com/elastic/kibana/commit/c655f50950693f3d0ec32dcf8167135d0c9781ec) |
| --- | --- | --- | --- |
| canvas | 1.4MB | +26.8KB | 1.4MB |
| spaces | 45.9KB | +122.0B | 45.8KB |
| total |  | +26.9KB |  |

#### page load bundle size
| id | [value](https://github.com/elastic/kibana/commit/fbdae3fa444b19f35edcb40e309c5bdc7ecc300c) | diff | [baseline](https://github.com/elastic/kibana/commit/c655f50950693f3d0ec32dcf8167135d0c9781ec) |
| --- | --- | --- | --- |
| canvas | 1.3MB | -3.0B | 1.3MB |
| spaces | 455.1KB | +42.2KB | 412.8KB |
| total |  | +42.2KB |  |

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/66935/","number":"66935","commit":"fbdae3fa444b19f35edcb40e309c5bdc7ecc300c"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/66968/)
* Commit: 6f8250c7d5cd9779146ba006d01080bf41e9c0a2

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/6f8250c7d5cd9779146ba006d01080bf41e9c0a2) | diff | [baseline](https://github.com/elastic/kibana/commit/c655f50950693f3d0ec32dcf8167135d0c9781ec) |
| --- | --- | --- | --- |
| apm | 1172 | +231 | 941 |
| canvas | 1270 | +67 | 1203 |
| spaces | 312 | +120 | 192 |
| total |  | +418 |  |

#### async chunks size
| id | [value](https://github.com/elastic/kibana/commit/6f8250c7d5cd9779146ba006d01080bf41e9c0a2) | diff | [baseline](https://github.com/elastic/kibana/commit/c655f50950693f3d0ec32dcf8167135d0c9781ec) |
| --- | --- | --- | --- |
| apm | 3.8MB | +72.3KB | 3.7MB |
| canvas | 1.4MB | +26.8KB | 1.4MB |
| spaces | 45.9KB | +122.0B | 45.8KB |
| total |  | +99.2KB |  |

#### page load bundle size
| id | [value](https://github.com/elastic/kibana/commit/6f8250c7d5cd9779146ba006d01080bf41e9c0a2) | diff | [baseline](https://github.com/elastic/kibana/commit/c655f50950693f3d0ec32dcf8167135d0c9781ec) |
| --- | --- | --- | --- |
| apm | 122.4KB | -135.6KB | 258.0KB |
| canvas | 1.3MB | -3.0B | 1.3MB |
| spaces | 455.1KB | +42.2KB | 412.8KB |
| total |  | -93.4KB |  |

### History
* :green_heart: [Build #66935](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/66935/) succeeded fbdae3fa444b19f35edcb40e309c5bdc7ecc300c

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/66935/","number":"66935","commit":"fbdae3fa444b19f35edcb40e309c5bdc7ecc300c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/66968/","number":"66968","commit":"6f8250c7d5cd9779146ba006d01080bf41e9c0a2"}]}
PIPELINE-->

watson reviewed (commented):
Looks good from a security perspective 👍 

I think it would be nice if we could add a linting rule in `.eslintrc.js` that forbids us to import directly from `lodash` within a `public` directory. I can easily see these types of imports sneaking in again, in which case all your good work is for nothing, so it would be good to guard against it.

Of the top of my head I think it would be something like this:

```js
    {
      files: ['**/public/**/*.{js,mjs,ts,tsx}'],
      rules: {
        'no-restricted-imports': [
          2,
          {
            paths: [
              {
                name: 'lodash',
                message: 'Please import specific method as a file instead, e.g. "lodash/get"',
              },
            ],
          },
        ],
      },
    },
```

elasticmachine commented:
Pinging @elastic/apm-ui (Team:apm)

brittanyjoiner15 commented:
[Per this discussion on Slack](https://elastic.slack.com/archives/C010ZGLH0SF/p1597054799049000), we're going to do this for the whole repo. I'll also work on adding the eslint rule above. Will be working on this this week!

brittanyjoiner15 commented:
@elasticmachine merge upstream

elasticmachine commented:
merge conflict between base and head

brittanyjoiner15 commented:
I've updated more files here :) There are still about 391 mentions of `} from 'lodash` that all contain `import _`, which from what I can tell means to pull the whole library. So are these instances in which we actually want to pull the whole library? Or should we remove the `_` and continue to just import the pieces in `{ }` ? 

Once we've clarified that, I'll make any edits, and then add in the `eslint` config, which i tried out and it worked just as @watson wrote above, but doesn't like all the remaining ones in the code, so i removed it so we can resolve this first.

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/67375/)
* Commit: 57b963e031de10f28657f224bfe9b7013783beac
* This comment will update when the build is complete

### Failed CI Steps
* [Setup Build Environment and Dependencies](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/67375/execution/node/105/log)
* [Setup Build Environment and Dependencies](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/67375/execution/node/111/log)

### History
* :green_heart: [Build #66968](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/66968/) succeeded 6f8250c7d5cd9779146ba006d01080bf41e9c0a2
* :green_heart: [Build #66935](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/66935/) succeeded fbdae3fa444b19f35edcb40e309c5bdc7ecc300c

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/66935/","number":"66935","commit":"fbdae3fa444b19f35edcb40e309c5bdc7ecc300c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/66968/","number":"66968","commit":"6f8250c7d5cd9779146ba006d01080bf41e9c0a2"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/67375/","number":"67375","commit":"57b963e031de10f28657f224bfe9b7013783beac"}]}
PIPELINE-->

TinaHeiligers reviewed (approved):
LGTM from a usage collection and telemetry pov. I also checked localization (they were left off of the list), that looks fine too.

walterra reviewed (approved):
ML/Transform changes LGTM.

watson commented:
> There are still about 391 mentions of } from 'lodash that all contain import _, which from what I can tell means to pull the whole library. So are these instances in which we actually want to pull the whole library? Or should we remove the _ and continue to just import the pieces in { } ?

I think those should technically also be converted, though I realise it's a bit more difficult. I actually thought I went over all those a few months back, but I can see now I only refactored the `* as _` variant: https://github.com/elastic/kibana/pull/67452/files#diff-6201e61afcaae1e0f3ace4517cba1046

In the process, I found that VSCode has a way to somewhat automate this - though you still have to do it on a per-file basis which sucks if there's 391 files you have to go through. However, I guess you only need to do this on files that will become part of the public bundle, and could skip the ones that are server-side only?

Here's how I did it in VSCode:

1. Convert `import _ from 'lodash'` to `import * as _ from 'lodash'` (I wish this step was redundant, but without it, VSCode doesn't seem to want to continue with the next step in the refactor)
1. Highlight the line containing the `import`. A lightbulb should now appear.
1. Click the lightbulb and select "Convert namespace import to named imports" in the dropdown:
![image](https://user-images.githubusercontent.com/10602/89872002-96482f80-dbb8-11ea-801f-ac4c32ec0da3.png)

afgomez reviewed (approved):
Infra changes LGTM

shahzad31 reviewed (approved):
Uptime changes LGTM !!

afharo commented on packages/kbn-analytics/babel.config.js:
```diff
@@ -1,32 +1,32 @@
-  },
-  ignore: ['**/*.test.ts', '**/*.test.tsx'],
-};
+/*
```
Not sure what changed in this file 😅

watson replied:
I think it's a CRLF -> LF replacement 🤔

afharo reviewed (approved):
Telemetry changes LGTM!

dgieselaar commented:
@brittanyjoiner15 I've looked at the changed files but I have some concerns about files being changed that should not be changed. How do you feel about:

- adding the lint rule as @watson suggested
- Running `eslint` locally, and getting a list of files
- Run the codemod on those files only
- Update the remaining files manually (not sure if this is reasonable)

dgieselaar commented:
@brittanyjoiner15 how did you run the codemod? I'm running into parsing issues:

> examples/alerting_example/public/alert_types/astros.tsx Transformation error (Unexpected token, expected "," (39:41))
SyntaxError: Unexpected token, expected "," (39:41)

dmlemeshko reviewed (approved):
Kibana QA changes look good.

jgowdyelastic commented:
@brittanyjoiner15 ML currently has a PR up to remove all `import _ from 'lodash'` 
https://github.com/elastic/kibana/pull/74742

mikecote reviewed (approved):
Alerting services code LGTM

brittanyjoiner15 commented:
@watson i was able to go back and change a lot of them! I realized a majority were happening in `.map` and `.json` files that were part of node modules, so we removed those, and then that only left about 21 that I went in and quickly fixed manually!



> > There are still about 391 mentions of } from 'lodash that all contain import _, which from what I can tell means to pull the whole library. So are these instances in which we actually want to pull the whole library? Or should we remove the _ and continue to just import the pieces in { } ?
> 
> I think those should technically also be converted, though I realise it's a bit more difficult. I actually thought I went over all those a few months back, but I can see now I only refactored the `* as _` variant: https://github.com/elastic/kibana/pull/67452/files#diff-6201e61afcaae1e0f3ace4517cba1046
> 
> In the process, I found that VSCode has a way to somewhat automate this - though you still have to do it on a per-file basis which sucks if there's 391 files you have to go through. However, I guess you only need to do this on files that will become part of the public bundle, and could skip the ones that are server-side only?
> 
> Here's how I did it in VSCode:
> 
> 1. Convert `import _ from 'lodash'` to `import * as _ from 'lodash'` (I wish this step was redundant, but without it, VSCode doesn't seem to want to continue with the next step in the refactor)
> 2. Highlight the line containing the `import`. A lightbulb should now appear.
> 3. Click the lightbulb and select "Convert namespace import to named imports" in the dropdown:
>    ![image](https://user-images.githubusercontent.com/10602/89872002-96482f80-dbb8-11ea-801f-ac4c32ec0da3.png)

brittanyjoiner15 commented:
@dgieselaar i'm good with that approach but might need some help - the code mod was a bit tricky and @smith helped me figure out all the ways of making it work. 

as for the code i ran it was `jscodeshift -t modular-lodash.js ../kibana/x-pack/plugins/apm --extensions=js,jsx,ts,tsx --parser=tsx --ignore-pattern=target`

> @brittanyjoiner15 I've looked at the changed files but I have some concerns about files being changed that should not be changed. How do you feel about:
> 
> * adding the lint rule as @watson suggested
> * Running `eslint` locally, and getting a list of files
> * Run the codemod on those files only
> * Update the remaining files manually (not sure if this is reasonable)

alisonelizabeth reviewed (approved):
ES UI changes LGTM.

mistic reviewed (approved):
Changes in files under operations team code owners LGTM. The other concerns I had were already raised both by @watson and @dgieselaar

poffdeluxe reviewed (approved):
Canvas changes LGTM

watson commented on src/core/public/chrome/doc_title/doc_title_service.ts:
```diff
@@ -17,7 +17,10 @@
 
-import { compact, flattenDeep, isString } from 'lodash';
+import compact from 'lodash/compact';
+
```
```suggestion
```

watson commented on src/core/public/plugins/plugins_service.test.ts:
```diff
@@ -17,7 +17,9 @@
 
-import { omit, pick } from 'lodash';
+import omit from 'lodash/omit';
+
```
```suggestion
```

watson commented on src/core/server/ui_settings/ui_settings_client.ts:
```diff
@@ -16,7 +16,9 @@
  */
-import { defaultsDeep, omit } from 'lodash';
+import defaultsDeep from 'lodash/defaultsDeep';
+
```
```suggestion
```

watson commented on src/legacy/server/kbn_server.js:
```diff
@@ -17,7 +17,11 @@
 
-import { constant, once, compact, flatten } from 'lodash';
+import constant from 'lodash/constant';
+
```
```suggestion
```

watson commented on src/legacy/server/status/lib/case_conversion.ts:
```diff
@@ -17,7 +17,9 @@
 
-import { mapKeys, snakeCase } from 'lodash';
+import mapKeys from 'lodash/mapKeys';
+
```
```suggestion
```

watson commented on src/legacy/server/status/wrap_auth_config.js:
```diff
@@ -17,7 +17,9 @@
 
-import { assign, identity } from 'lodash';
+import assign from 'lodash/assign';
+
```
```suggestion
```

watson commented on src/legacy/ui/public/directives/render_directive/apply_scope_bindings.js:
```diff
@@ -17,7 +17,9 @@
 
-import { forOwn, noop } from 'lodash';
+import forOwn from 'lodash/forOwn';
+
```
```suggestion
```

watson commented on src/legacy/ui/public/indexed_array/helpers/organize_by.ts:
```diff
@@ -17,7 +17,9 @@
 
-import { each, isFunction } from 'lodash';
+import each from 'lodash/each';
+
```
```suggestion
```

watson commented on src/legacy/ui/public/routes/breadcrumbs.js:
```diff
@@ -17,7 +17,9 @@
 
-import { trim, startCase } from 'lodash';
+import trim from 'lodash/trim';
+
```
```suggestion
```

watson commented on src/legacy/ui/public/routes/route_manager.js:
```diff
@@ -17,7 +17,10 @@
 
-import { cloneDeep, defaultsDeep, wrap } from 'lodash';
+import cloneDeep from 'lodash/cloneDeep';
+
```
```suggestion
```

watson commented on src/plugins/console/public/application/models/legacy_core_editor/smart_resize.ts:
```diff
@@ -17,7 +17,9 @@
 
-import { get, throttle } from 'lodash';
+import get from 'lodash/get';
+
```
```suggestion
```

watson commented on src/plugins/console/public/services/storage.ts:
```diff
@@ -17,7 +17,10 @@
 
-import { transform, keys, startsWith } from 'lodash';
+import transform from 'lodash/transform';
+
```
```suggestion
```

watson commented on src/plugins/dashboard/server/saved_objects/dashboard_migrations.ts:
```diff
@@ -17,7 +17,9 @@
 
-import { get, flow } from 'lodash';
+import get from 'lodash/get';
+
```
```suggestion
```

watson commented on src/plugins/data/common/field_formats/field_formats_registry.ts:
```diff
@@ -18,7 +18,11 @@
 // eslint-disable-next-line max-classes-per-file
-import { forOwn, isFunction, memoize, identity } from 'lodash';
+import forOwn from 'lodash/forOwn';
+
```
```suggestion
```

watson commented on src/plugins/data/common/index_patterns/index_patterns/index_pattern.test.ts:
```diff
@@ -17,7 +17,11 @@
 
-import { defaults, map, last, get } from 'lodash';
+import defaults from 'lodash/defaults';
+
```
```suggestion
```

watson commented on src/plugins/data/server/index_patterns/fetcher/lib/field_capabilities/field_capabilities.ts:
```diff
@@ -17,7 +17,10 @@
 
-import { defaults, keyBy, sortBy } from 'lodash';
+import defaults from 'lodash/defaults';
+
```
```suggestion
```

watson commented on src/plugins/expressions/common/expression_types/specs/datatable.ts:
```diff
@@ -17,7 +17,10 @@
 
-import { map, pick, zipObject } from 'lodash';
+import map from 'lodash/map';
+
```
```suggestion
```

watson commented on src/plugins/kibana_utils/public/field_wildcard/field_wildcard.ts:
```diff
@@ -17,7 +17,9 @@
 
-import { escapeRegExp, memoize } from 'lodash';
+import escapeRegExp from 'lodash/escapeRegExp';
+
```
```suggestion
```

watson commented on src/plugins/kibana_utils/public/state_management/utils/diff_object.ts:
```diff
@@ -16,7 +16,16 @@
  */
-import { keys, isFunction, difference, filter, union, pick, each, assign, isEqual } from 'lodash';
+import keys from 'lodash/keys';
+
```
```suggestion
```

watson commented on src/plugins/saved_objects_management/public/lib/case_conversion.ts:
```diff
@@ -17,7 +17,9 @@
 
-import { mapKeys, camelCase } from 'lodash';
+import mapKeys from 'lodash/mapKeys';
+
```
```suggestion
```

watson commented on src/plugins/vis_default_editor/public/components/agg_params_helper.ts:
```diff
@@ -17,7 +17,9 @@
 
-import { get, isEmpty } from 'lodash';
+import get from 'lodash/get';
+
```
```suggestion
```

watson commented on src/plugins/vis_type_timeseries/common/get_last_value.js:
```diff
@@ -17,7 +17,9 @@
 
-import { isArray, last } from 'lodash';
+import isArray from 'lodash/isArray';
+
```
```suggestion
```

watson commented on src/plugins/vis_type_timeseries/public/application/components/lib/charts.js:
```diff
@@ -17,6 +17,10 @@
 
-import { uniq, map, size, flow } from 'lodash';
+import uniq from 'lodash/uniq';
+
```
```suggestion
```

watson commented on src/plugins/vis_type_timeseries/server/lib/vis_data/response_processors/annotations/buckets.js:
```diff
@@ -17,7 +17,9 @@
 
-import { get, isEmpty } from 'lodash';
+import get from 'lodash/get';
+
```
```suggestion
```

watson commented on src/plugins/visualize/public/application/utils/create_visualize_app_state.ts:
```diff
@@ -17,7 +17,10 @@
 
-import { isFunction, omitBy, union } from 'lodash';
+import isFunction from 'lodash/isFunction';
+
```
```suggestion
```

watson commented on x-pack/plugins/canvas/canvas_plugin_src/renderers/plot/plugins/size.js:
```diff
@@ -4,7 +4,11 @@
 
-import { min, max, map, flatten } from 'lodash';
+import min from 'lodash/min';
+
```
```suggestion
```

watson commented on x-pack/plugins/canvas/common/lib/pivot_object_array.ts:
```diff
@@ -4,7 +4,9 @@
 
-import { map, zipObject } from 'lodash';
+import map from 'lodash/map';
+
```
```suggestion
```

watson commented on x-pack/plugins/infra/common/ecs_allowed_list.ts:
```diff
@@ -3,7 +3,9 @@
  */
-import { first, memoize } from 'lodash';
+import first from 'lodash/first';
+
```
```suggestion
```

watson commented on x-pack/plugins/logstash/public/lib/get_search_value/get_search_value.js:
```diff
@@ -4,7 +4,9 @@
 
-import { pick, values } from 'lodash';
+import pick from 'lodash/pick';
+
```
```suggestion
```

watson commented on x-pack/plugins/logstash/public/models/pipeline/pipeline.js:
```diff
@@ -4,7 +4,11 @@
 
-import { get, omit, isEqual, defaultsDeep } from 'lodash';
+import get from 'lodash/get';
+
```
```suggestion
```

watson commented on x-pack/plugins/ml/public/application/timeseriesexplorer/timeseriesexplorer_utils/validate_job_selection.ts:
```diff
@@ -4,7 +4,9 @@
 
-import { difference, without } from 'lodash';
+import difference from 'lodash/difference';
+
```
```suggestion
```

watson commented on x-pack/plugins/monitoring/public/components/chart/get_last_value.js:
```diff
@@ -4,7 +4,10 @@
 
-import { isNumber, get, last } from 'lodash';
+import isNumber from 'lodash/isNumber';
+
```
```suggestion
```

watson commented on x-pack/plugins/monitoring/server/kibana_monitoring/bulk_uploader.js:
```diff
@@ -4,7 +4,10 @@
 
-import { defaultsDeep, uniq, compact } from 'lodash';
+import defaultsDeep from 'lodash/defaultsDeep';
+
```
```suggestion
```

watson commented on x-pack/plugins/monitoring/server/lib/beats/_beats_stats.js:
```diff
@@ -4,7 +4,9 @@
 
-import { upperFirst, get } from 'lodash';
+import upperFirst from 'lodash/upperFirst';
+
```
```suggestion
```

watson commented on x-pack/plugins/monitoring/server/lib/ccs_utils.js:
```diff
@@ -3,7 +3,9 @@
  */
-import { isFunction, get } from 'lodash';
+import isFunction from 'lodash/isFunction';
+
```
```suggestion
```

watson commented on x-pack/plugins/monitoring/server/lib/elasticsearch/nodes/calculate_node_type.js:
```diff
@@ -11,7 +11,9 @@
  */
-import { includes, isUndefined } from 'lodash';
+import includes from 'lodash/includes';
+
```
```suggestion
```

watson commented on x-pack/plugins/rollup/public/crud_app/sections/job_create/steps_config/index.js:
```diff
@@ -4,7 +4,10 @@
 
-import { cloneDeep, get, pick } from 'lodash';
+import cloneDeep from 'lodash/cloneDeep';
+
```
```suggestion
```

watson commented on x-pack/plugins/watcher/public/application/lib/get_search_value.ts:
```diff
@@ -4,7 +4,9 @@
 
-import { pick, values } from 'lodash';
+import pick from 'lodash/pick';
+
```
```suggestion
```

watson commented on x-pack/plugins/watcher/server/models/execute_details/execute_details.js:
```diff
@@ -4,7 +4,9 @@
 
-import { omitBy, isUndefined } from 'lodash';
+import omitBy from 'lodash/omitBy';
+
```
```suggestion
```

watson reviewed (commented):
Just went over all the files. A few of them had an extra blank line added in the middle of the `lodash` imports (just a tiny nit, sorry). But I've made a suggested change which hopefully should be simple to apply 😅 

A few files have been edited in this PR that didn't have any `lodash` imports (I presume it's a prettier update). If you want to keep the PR clean it might be best to leave those our of the PR - but I'll leave it up to you. I've made a list of them which I can share if you like.

I still think it's ok to use `{...}` style imports server-side, in tests etc. Did you consider only doing this for code that's part of the frontend bundle?

dgieselaar commented:
@watson should it be `'**/{public,common}/**/*.{js,mjs,ts,tsx}' then?

jfsiii reviewed (approved):
LGTM for the 1 Ingest Manager change in `x-pack/plugins/ingest_manager/server/services/agent_config.ts`

watson commented:
@dgieselaar yes + possibly everything under the `packages` directory as well, since those might be used client side also right? Sorry I can't come up with a complete list off the top of my head. My knowledge of how the frontend bundle is close to non-existent :D

dgieselaar commented:
@brittanyjoiner15 After trying out a couple of things I am leaning towards the following strategy:

**1. Manually replace all occurrences of `chain` with cherrypick imports.**

I don't think `chain` will work with cherry-picked imports. [This comment explains it a bit](https://github.com/lodash/lodash/issues/4275#issuecomment-485048814). We might be better off by making that change in a separate PR as we cannot automate it (to my knowledge) and it needs a more focused review (from the teams that are using it). We can just search for `_.chain` in the repository. At the time of writing, I'm seeing 15 results in 11 files.

Example:

```ts
import _ from 'lodash';

export default function unzipPairs(timeValObject) {
  const paired = _.chain(timeValObject)
    .toPairs()
    .map(function (point) {
      return [parseInt(point[0], 10), point[1]];
    })
    .sortBy(function (point) {
      return point[0];
    })
    .value();
  return paired;
}
```

Should be replaced by:

```ts
import toPairs from 'lodash/toPairs';
import sortBy from 'lodash/sortBy';

export default function unzipPairs(timeValObject) {

  const paired = toPairs(timeValObject)
    .map(function (point) {
      return [parseInt(point[0], 10), point[1]];
    });

  const sorted = sortBy(paired, function (point) {
    return point[0];
  });

  return sorted;
}
```

**2. Convert the remaining default imports of `lodash` to namespace imports, per @watson's suggestion.**

You can run a find-and-replace on `import _ from 'lodash'` to `import * as _ from 'lodash'`. I didn't find any other patterns (like `import _, { map } from 'lodash'`). I've looked around for a codemod but I think Thomas' VSCode refactor approach is our best bet here.

**3. Convert named imports for lodash with a codemod.**

I think we should run this codemod _only_ on files that we want to change. My suggestion would be to run ESLint with the updated ruleset (again per @watson's suggestion) after completing step 1 and 2, get a list of files that create warnings for the added rule, and pass those files explicitly to `jscodeshift`.

@watson @elastic/kibana-operations WDYT? @brittanyjoiner15 happy to Zoom about it later this week if you have the time (or maybe someone else can help out if timezones are an issue).

watson commented:
@dgieselaar sounds good 👍 If there's anything I can help with, I can jump on a call as well.

chrisronline commented on src/legacy/server/logging/log_format.js:
```diff
@@ -19,7 +19,8 @@
 import moment from 'moment-timezone';
-import { get, _ } from 'lodash';
+import get from 'lodash/get';
+import _ from 'lodash/_';
```
Yea I'm seeing an error here trying to run this PR
```
Error: Cannot find module 'lodash/_'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
```

chrisronline commented on src/plugins/data/public/search/aggs/buckets/lib/time_buckets/time_buckets.ts:
```diff
@@ -17,7 +17,10 @@
 
-import { isString, isObject as isObjectLodash, isPlainObject, sortBy } from 'lodash';
+import isString from 'lodash/isString';
+import isObject from 'lodash/isObject';
```
This should be `isObjectLodash`

tylersmalley replied:
Added as a suggestion to make it easy to apply.

tylersmalley commented on src/plugins/data/public/search/aggs/buckets/lib/time_buckets/time_buckets.ts:
```diff
@@ -17,7 +17,10 @@
 
-import { isString, isObject as isObjectLodash, isPlainObject, sortBy } from 'lodash';
+import isString from 'lodash/isString';
+import isObject from 'lodash/isObject';
```
```suggestion
import isObject as isObjectLodash from 'lodash/isObject';
```

chrisronline commented on x-pack/plugins/monitoring/server/lib/kibana/get_kibanas_for_clusters.js:
```diff
@@ -5,7 +5,9 @@
 
 import Bluebird from 'bluebird';
-import { chain, find, get } from 'lodash';
+import chain from 'lodash/chain';
```
I'm seeing this error with this change:
```
server    log   [13:32:08.847] [error][monitoring][monitoring][plugins] TypeError: (0 , _chain.default)(...).map is not a function
    at callWithRequest.then.result (/Users/chris/dev/repos/kibana/x-pack/plugins/monitoring/server/lib/kibana/get_kibanas_for_clusters.js:183:12)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```

watson replied:
Is this related to point no 1 in @dgieselaar comment?: https://github.com/elastic/kibana/pull/74539#issuecomment-672808058

chrisronline replied:
Yes, thank you! Same thing indeed

pgayvallet commented:
Quick question here, but why not leverage webpack/babel to do this? There are already existing modules to do it at compile time, such as https://github.com/lodash/babel-plugin-lodash?

watson commented:
@pgayvallet that's a really good idea. It would mean that the development workflow isn't impacted (as it would be somewhat by the eslint rule). I didn't know about this babel plugin.

dgieselaar commented:
@pgayvallet @watson SIEM is using that plugin IIRC, but I think it was @tylersmalley's preference to just have explicit cherry-picked imports - at least that's what I assumed based off a Slack conversation. @tylersmalley thoughts?

tylersmalley commented:
@pgayvallet @watson @dgieselaar

For SIEM, that was done [here](https://github.com/elastic/kibana/blob/master/packages/kbn-babel-preset/webpack_preset.js#L43-L60).

I personally prefer we write the code as we expect it to behave instead of relying on transformations. It really makes debugging a nightmare. Do you all disagree?

This change might need to be broken up and a few plugins at a time. We have done this in the past and added exceptions to the ESLint rule and pointed to issues for the plugins to remove the exception and become compliant.

mistic commented:
I agree with @tylersmalley on this one. Although I'm not sure what will be the penalty impact of transform this at compile time, our lodash usage is really wide and I would prefer to have the code explicitly behave as expected. Another thing to consider is that in a future lodash major upgrade it would be easier to find the places with lodash usages we need to convert because we can just search for `lodash/{function_name}`. Finally, I'm not sure if the plugin is working as expected right now: https://github.com/lodash/babel-plugin-lodash/issues/236. If we endup deciding to go through the transformation route, `babel-plugin-transform-imports` probably would be a better alternative as while it does not remap default imports it works as expected over named imports.

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70030/)
* Commit: e620a163676577f55ba7d35f937a97ea11a1108c
* This comment will update when the build is complete

### Failed CI Steps
* [Setup Build Environment and Dependencies](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70030/execution/node/105/log)
* [Setup Build Environment and Dependencies](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70030/execution/node/111/log)

### History
* :broken_heart: [Build #67375](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/67375/) failed 57b963e031de10f28657f224bfe9b7013783beac
* :green_heart: [Build #66968](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/66968/) succeeded 6f8250c7d5cd9779146ba006d01080bf41e9c0a2
* :green_heart: [Build #66935](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/66935/) succeeded fbdae3fa444b19f35edcb40e309c5bdc7ecc300c

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/66935/","number":"66935","commit":"fbdae3fa444b19f35edcb40e309c5bdc7ecc300c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/66968/","number":"66968","commit":"6f8250c7d5cd9779146ba006d01080bf41e9c0a2"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/67375/","number":"67375","commit":"57b963e031de10f28657f224bfe9b7013783beac"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70030/","number":"70030","commit":"e620a163676577f55ba7d35f937a97ea11a1108c"}]}
PIPELINE-->

brittanyjoiner15 commented:
Nathan and I are doing this...

In summary, what we need to watch out for:
- [ ] Make sure no extra new lines are added
- [ ] Only edit in `/public,/common`, and `/packages`


to do: 
- [x] Revert to master
- [ ] do find `import _, (\{ [a-zA-Z\s,]+ \}) from 'lodash'` and replace with `import $1 from 'lodash'`
- [ ] write and run [ESlint rule](https://github.com/elastic/kibana/pull/74539#pullrequestreview-464103849) to get list of files
- [ ] run codemod on tht list of files

dgieselaar commented:
Please keep _.chain in mind 😀 I would prefer for it to be a separate PR.
As we cannot automate it I feel we need to give it a little more attention
than the other automated steps.

Op ma 24 aug. 2020 18:09 schreef Brittany Joiner <notifications@github.com>:

> Nathan and I are doing this...
>
> In summary, what we need to watch out for:
>
>    - Make sure no extra new lines are added
>    - Only edit in /public,/common, and /packages
>
> to do:
>
>    - Revert to master
>    - do find import _, (\{ [a-zA-Z\s,]+ \}) from 'lodash' and replace
>    with import $1 from 'lodash'
>    - write and run ESlint rule
>    <https://github.com/elastic/kibana/pull/74539#pullrequestreview-464103849>
>    to get list of files
>    - run codemod on tht list of files
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/kibana/pull/74539#issuecomment-679221696>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACWDXHTIAZJCVZMS5RSE33SCKGCZANCNFSM4PWVSFNA>
> .
>

dgieselaar commented:
Also, disabling the prettier rule when running eslint should give you a
nice speed boost of 75%. There are some other rules that could save you
some time too, I posted the list in the Kibana ops channel.

Op ma 24 aug. 2020 19:11 schreef Dario Gieselaar <dario.gieselaar@elastic.co
>:

> Please keep _.chain in mind 😀 I would prefer for it to be a separate PR.
> As we cannot automate it I feel we need to give it a little more attention
> than the other automated steps.
>
> Op ma 24 aug. 2020 18:09 schreef Brittany Joiner <notifications@github.com
> >:
>
>> Nathan and I are doing this...
>>
>> In summary, what we need to watch out for:
>>
>>    - Make sure no extra new lines are added
>>    - Only edit in /public,/common, and /packages
>>
>> to do:
>>
>>    - Revert to master
>>    - do find import _, (\{ [a-zA-Z\s,]+ \}) from 'lodash' and replace
>>    with import $1 from 'lodash'
>>    - write and run ESlint rule
>>    <https://github.com/elastic/kibana/pull/74539#pullrequestreview-464103849>
>>    to get list of files
>>    - run codemod on tht list of files
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/elastic/kibana/pull/74539#issuecomment-679221696>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AACWDXHTIAZJCVZMS5RSE33SCKGCZANCNFSM4PWVSFNA>
>> .
>>
>

brittanyjoiner15 committed:
changed lodash

brittanyjoiner15 committed:
prettier

brittanyjoiner15 committed:
removed spaces

brittanyjoiner15 committed:
fixed aliases

brittanyjoiner15 committed:
fixed prettier rule

brittanyjoiner15 commented:
@elasticmachine merge upstream

elasticmachine commented:
merge conflict between base and head

brittanyjoiner15 commented:
We're close! Just need to fetch the most recent and do this again to avoid conflicts.

- [ ] turn off prettier
- [ ] get list of files ( $`yarn lint:es --format unix > output` and then `cat Output | cut -d ":" -f 1 > listOfFiles` )
- [ ] run the codemod (`jscodeshift --stdin --extensions=js,jsx,ts,tsx -t modular-lodash.js --parser=tsx < listOfFiles`)
- [ ] run the regex to fix new line spaces (find: `import(.*)lodash(.*)\n\nimport(.*)` replace `import$1lodash$2\nimport$3`
- [ ] run the regex to fix aliases (`git diff upstream/master | code - ` and search for `import.*\sas\s.*lodash` )
- [ ] turn prettier back on

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72750/)
* Commit: 6c6b1cf2e2f182e8499c2a4147650373295e64ff
* This comment will update when the build is complete

### Failed CI Steps
* [Check Telemetry Schema](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72750/execution/node/354/log)

### History
* :broken_heart: [Build #70030](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70030/) failed e620a163676577f55ba7d35f937a97ea11a1108c
* :broken_heart: [Build #67375](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/67375/) failed 57b963e031de10f28657f224bfe9b7013783beac
* :green_heart: [Build #66968](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/66968/) succeeded 6f8250c7d5cd9779146ba006d01080bf41e9c0a2
* :green_heart: [Build #66935](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/66935/) succeeded fbdae3fa444b19f35edcb40e309c5bdc7ecc300c

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/66935/","number":"66935","commit":"fbdae3fa444b19f35edcb40e309c5bdc7ecc300c"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/66968/","number":"66968","commit":"6f8250c7d5cd9779146ba006d01080bf41e9c0a2"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/67375/","number":"67375","commit":"57b963e031de10f28657f224bfe9b7013783beac"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/70030/","number":"70030","commit":"e620a163676577f55ba7d35f937a97ea11a1108c"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/72750/","number":"72750","commit":"6c6b1cf2e2f182e8499c2a4147650373295e64ff"}]}
PIPELINE-->

stacey-gammon commented:
is this PR old and should be closed? 7.9.2 FF is today... also this doesn't seem like it should go into a patch release.  Going to remove the label!

smith commented:
> is this PR old and should be closed? 7.9.2 FF is today... also this doesn't seem like it should go into a patch release. Going to remove the label!

Thanks for removing the label. It's still in progress but it's a big hairy change.

smith commented:
Closing in favor of #78156.
